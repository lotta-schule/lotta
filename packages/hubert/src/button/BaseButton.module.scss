.root {
  position: relative;
  border: 1px solid rgba(var(--lotta-primary-color), 1);
  color: rgba(var(--lotta-primary-color), 1);
  background: rgba(var(--lotta-box-background-color), 1);
  border-radius: 0.5em;
  padding: 0.7em;
  display: flex;
  text-transform: uppercase;
  letter-spacing: 1px;
  text-decoration: none;
  font-size: 0.8rem;
  justify-content: space-evenly;
  font-family: var(--lotta-text-font-family);
  &:not([disabled]) {
    cursor: pointer;
    &:hover,
    &:active {
      transition: linear;
      transition-duration: 0.2s;
      background-color: rgba(var(--lotta-primary-color), 0.1);
    }
  }

  &[disabled] {
    color: rgba(var(--lotta-disabled-color), 1);
    border-color: rgba(var(--lotta-disabled-color), 1);
  }

  &.selected {
    border: 1px solid rgba(var(--lotta-primary-color), 1);
    background-color: rgba(var(--lotta-primary-color), 0.2);
    filter: brightness(90%);

    &:hover {
      filter: brightness(90%);
    }
    &:active {
      filter: brightness(70%);
    }
  }

  &.fullWidth {
    width: 100%;
  }

  /**
      * Custom styles for different variants
      */
  &[data-variant='fill'] {
    background-color: rgba(var(--lotta-primary-color), 1);
    color: rgba(var(--lotta-contrast-text-color), 1);
    border: none;
    &:hover {
      transition: linear;
      transition-duration: 0.2s;
      background: rgba(var(--lotta-primary-color), 1);
      box-shadow: inset 0 0 20px #00000024;
    }
    &:active {
      background: rgba(var(--lotta-primary-color), 1);
      box-shadow: inset 0 0 20px #0000004f;
      filter: none;
    }
  }
  &[data-variant='borderless'] {
    border: none;
  }
  &[data-variant='error'] {
    background-color: rgba(var(--lotta-error-color), 1);
    color: #fff;
    border: none;
    &:hover {
      transition: linear;
      transition-duration: 0.2s;
      background-color: rgba(var(--lotta-error-color), 1);
      box-shadow: inset 0 0 20px #00000024;
    }
    &:active {
      background-color: rgba(var(--lotta-primary-color), 1);
      box-shadow: inset 0 0 20px #0000004f;
      filter: none;
    }
  }
}
