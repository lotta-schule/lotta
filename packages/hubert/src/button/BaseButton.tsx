import * as React from 'react';
import clsx from 'clsx';

import styles from './BaseButton.module.scss';

export type BaseButtonProps = {
  /**
   * Wether the button is currently 'active' or 'selected'
   */
  selected?: boolean;

  /**
   * Chose from different styles
   */
  variant?: 'default' | 'fill' | 'borderless' | 'error';

  /**
   * Wether the button should have a width of 100%
   */
  fullWidth?: boolean;

  /**
   * CSS Classname
   */
  className?: string;

  /**
   * CSS Styles
   */
  style?: React.CSSProperties;

  /**
   * Content to show on the button
   */
  children?: any;
} & React.HTMLProps<HTMLButtonElement>;

/**
 * Primary UI shared for userAvatar interaction
 */
export const BaseButton = React.forwardRef<any, BaseButtonProps>(
  (
    {
      children,
      style,
      className,
      variant = 'default',
      fullWidth = false,
      selected = false,
      ...props
    },
    ref
  ) => {
    const ComponentClass = props.href ? 'a' : ('button' as any);
    const selectedAriaAttribute =
      props.role && ['gridcell', 'option', 'row', 'tab'].includes(props.role)
        ? 'aria-selected'
        : 'aria-current';
    return React.createElement(
      ComponentClass,
      {
        ref,
        type: props.type ?? 'button',
        role: 'button',
        style,
        'data-variant': variant,
        [selectedAriaAttribute]: selected,
        className: clsx(styles.root, className, {
          [styles.selected]: selected,
          [styles.fullWidth]: fullWidth,
        }),
        ...props,
      },
      children
    );
  }
);
BaseButton.displayName = 'BaseButton';
