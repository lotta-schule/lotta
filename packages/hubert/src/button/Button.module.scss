.root {
  color: rgba(var(--lotta-primary-color), 1);
  font-weight: 200;
  padding: 0.4em 0.8em;
  align-items: center;

  &:not(.small) {
    min-height: 2.8em;
  }

  .icon {
    display: flex;
  }

  .label {
    margin: auto;
    display: contents;
  }

  &.onlyIcon {
    &:not(.grouped) {
      border-radius: 50%;
      border-color: transparent;
      padding: 0.4em;

      &:not([data-varaint='fill']) {
        &:hover {
          background-color: rgba(var(--lotta-primary-color), 0.2);
        }
      }
    }
  }
}
