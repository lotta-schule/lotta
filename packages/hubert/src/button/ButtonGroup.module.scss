.root {
  display: inline-flex;
  flex-direction: row;
  justify-content: stretch;
  button {
    color: rgba(var(--lotta-primary-color), 1);
    background-color: #fff;
    display: inline-block;
    &:not(:only-child) {
      &:not(:first-of-type):not(:last-of-type):not(.isFirstButton):not(.isLastButton) {
        border-radius: 0;
        border-right: 0;
      }
      &:first-of-type:not(.isLastButton),
      &.isFirstButton {
        border-top-right-radius: 0;
        border-bottom-right-radius: 0;
        border-right: 0;
      }
      &:last-of-type:not(.isFirstButton),
      &.isLastButton {
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
      }
    }
    &.isFirstButton {
      border-top-right-radius: 0;
      border-bottom-right-radius: 0;
      border-right: 0;
    }
    &.isLastButton {
      border-top-left-radius: 0;
      border-bottom-left-radius: 0;
    }
    &.fullWidth {
      flex-grow: 1;
    }
    svg {
      font-size: 1.3rem;
    }
  }
  &.fullWidth {
    width: 100%;
  }
}
