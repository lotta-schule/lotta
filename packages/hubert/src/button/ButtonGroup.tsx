'use client';

import * as React from 'react';
import { ButtonGroupContextProvider } from './ButtonGroupContext';
import clsx from 'clsx';

import styles from './ButtonGroup.module.scss';

export interface ButtonGroupProps {
  fullWidth?: boolean;
  className?: string;
  style?: React.CSSProperties;
  children?: React.ReactNode | React.ReactNode[];
}

export const ButtonGroupHelperClasses = {
  isFirstButton: styles.isFirstButton,
  isLastButton: styles.isLastButton,
};

export const ButtonGroup = ({
  fullWidth,
  className,
  style,
  children,
}: ButtonGroupProps) => {
  return (
    <ButtonGroupContextProvider>
      <div
        role={'group'}
        style={style}
        className={clsx(styles.root, className, { fullWidth })}
      >
        {children}
      </div>
    </ButtonGroupContextProvider>
  );
};
