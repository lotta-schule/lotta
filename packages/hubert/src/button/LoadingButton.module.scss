.root {
  .circularProgress {
    transition: all 0.35s ease-in;
    position: absolute;
    top: var(--lotta-spacing);
    left: calc(var(--lotta-spacing) * 2);
    opacity: 0;
  }
  .label {
    transition: all 0.25s ease-in;
    opacity: 1;
    margin: 0
      calc(var(--lotta-spacing) + calc(calc(1em + var(--lotta-spacing)) / 2));
  }

  &.isLoading {
    .circularProgress {
      opacity: 1;
    }
    .label {
      margin: 0 0 0
        calc(calc(2 * var(--lotta-spacing)) + calc(1em + var(--lotta-spacing)));
    }
  }
}
