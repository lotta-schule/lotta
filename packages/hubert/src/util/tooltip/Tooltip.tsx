import * as React from 'react';
import { useTooltipTrigger } from '@react-aria/tooltip';
import { useTooltipTriggerState } from '@react-stately/tooltip';
import { mergeProps } from '@react-aria/utils';
import { TooltipOverlay } from './TooltipOverlay';

import styles from './Tooltip.module.scss';

export interface TooltipProps extends React.HTMLAttributes<HTMLDivElement> {
  children: React.ReactElement;
  label?: string | React.ReactElement<{}>;
  /**
   * The delay time for the tooltip to show up
   * @default 750
   */
  delay?: number;
}

export const Tooltip = React.forwardRef(
  (
    { children, className, label, delay = 750, ...props }: TooltipProps,
    forwardedRef: React.Ref<HTMLDivElement | null>
  ) => {
    const ref = React.useRef<HTMLDivElement>(null);
    React.useImperativeHandle(forwardedRef, () => ref.current);

    const isDisabled = ref.current?.getAttribute('disabled') !== null;

    const state = useTooltipTriggerState({ delay, isDisabled });

    const { triggerProps, tooltipProps } = useTooltipTrigger(
      { delay, isDisabled },
      state,
      ref
    );

    return (
      <div className={styles.root}>
        <div className={styles.trigger} ref={ref} {...triggerProps}>
          {children}
        </div>
        <TooltipOverlay
          className={className}
          state={state}
          label={label ?? children}
          {...mergeProps(props, tooltipProps)}
        />
      </div>
    );
  }
);
Tooltip.displayName = 'Tooltip';
