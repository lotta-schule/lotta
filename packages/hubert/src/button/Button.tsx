import * as React from 'react';
import { BaseButton, BaseButtonProps } from './BaseButton';
import { useButtonGroupContext } from './ButtonGroupContext';
import clsx from 'clsx';

import styles from './Button.module.scss';

export type ButtonProps = {
  /**
   * Button contents
   */
  label?: string;

  /**
   * Icon to show on the button side
   */
  icon?: React.ReactElement;

  /**
   * Disable setting a minimal height
   */
  small?: boolean;

  /**
   * Force style for button with only icon and without label,
   * even if child is found
   */
  onlyIcon?: boolean;

  /**
   * Add custom claaes to the element tree
   **/
  classes?: {
    root?: string;
    onlyIcon?: string;
    icon?: string;
    label?: string;
  };
} & BaseButtonProps;

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (
    { icon, label, onlyIcon, small, children, classes, className, ...props },
    ref
  ) => {
    const showOnlyIcon = onlyIcon || (icon && !(label || children));
    const { grouped } = useButtonGroupContext();

    return (
      <BaseButton
        {...props}
        ref={ref}
        className={clsx(
          styles.root,
          className,
          classes?.root,
          showOnlyIcon && [styles.onlyIcon, classes?.onlyIcon],
          {
            [styles.small]: small,
            [styles.grouped]: grouped,
          }
        )}
      >
        {icon && (
          <span className={clsx(styles.icon, classes?.icon)}>{icon}</span>
        )}
        {(label ?? children) && (
          <span className={clsx(styles.label, classes?.label)}>
            {label ?? children}
          </span>
        )}
      </BaseButton>
    );
  }
);
Button.displayName = 'Button';
