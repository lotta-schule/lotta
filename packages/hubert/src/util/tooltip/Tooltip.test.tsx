import * as React from 'react';
import { render } from '@testing-library/react';
import { Button } from '../../button/Button';
import { Tooltip } from './Tooltip';
import userEvent from '@testing-library/user-event';

describe('util/Tooltip', () => {
  it('should render', () => {
    const screen = render(
      <Tooltip label="Test">
        <Button>Test</Button>
      </Tooltip>
    );
    expect(screen.container).toMatchSnapshot();
  });

  it('should keep a button clickable', async () => {
    const fireEvent = userEvent.setup();
    const onClick = jest.fn();
    const screen = render(
      <Tooltip label="Test">
        <Button onClick={onClick}>Test</Button>
      </Tooltip>
    );
    await fireEvent.click(screen.getByRole('button'));
    expect(onClick).toHaveBeenCalled();
  });
});
