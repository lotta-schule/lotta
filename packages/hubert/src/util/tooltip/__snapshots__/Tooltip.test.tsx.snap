// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util/Tooltip should render 1`] = `
<div>
  <div
    class="root"
  >
    <div
      class="trigger"
    >
      <button
        aria-current="false"
        class="root root"
        data-variant="default"
        role="button"
        type="button"
      >
        <span
          class="label"
        >
          Test
        </span>
      </button>
    </div>
  </div>
</div>
`;
